pywhy_graphs/algorithms/generic.py

```diff
@@ -855,3 +856,99 @@ def all_vstructures(G: nx.DiGraph, as_edges: bool = False):
                 else:
                     vstructs.add((p1, node, p2))  # type: ignore
     return vstructs
+
+
+def check_back_arrow(G: ADMG, X, Y: set):
+
+    out = set()
+
+    for elem in Y:
+        if not (
+            G.has_edge(X, elem, G.bidirected_edge_name) or G.has_edge(elem, X, G.directed_edge_name)
+        ):
+            out.update(elem)
+
+    return out
+
+
+def get_X_neighbors(G, X: set):
+
+    out = []
+
+    for elem in X:
+        elem_neighbors = set(G.neighbors(elem))
+        elem_possible_neighbors = check_back_arrow(G, elem, elem_neighbors)
+        to_remove = X.intersection(elem_possible_neighbors)
+        elem_neighbors = elem_possible_neighbors - to_remove
+
+        if len(elem_neighbors) != 0:
+            for nbh in elem_neighbors:
+                temp = dict()
+                temp[0] = elem
+                temp[1] = nbh
+                out.append(temp)
+    return out
+
+
+def recursively_find_pd_paths(G, X, paths, Y):
+
+    counter = 0
+    new_paths = []
+
+    for i in range(len(paths)):
+        cur_elem = paths[i][list(paths[i].keys())[-1]]
+
+        if cur_elem in Y:
+            new_paths.append(paths[i])
+            continue
+
+        nbr_temp = G.neighbors(cur_elem)
+        nbr_possible = check_back_arrow(G, cur_elem, nbr_temp)
+
+        if len(nbr_possible) == 0:
+            new_paths.append(paths[i].copy())
+
+        possible_end = nbr_possible.intersection(Y)
+
+        if len(possible_end) != 0:
+            for elem in possible_end:
+                temp_path = paths[i].copy()
+                temp_path[len(temp_path)] = elem
+                new_paths.append(temp_path)
+
+        remaining_nodes = nbr_possible - possible_end
+        remaining_nodes = (
+            remaining_nodes
+            - remaining_nodes.intersection(paths[i].values())
+            - remaining_nodes.intersection(X)
+        )
+
+        temp_arr = []
+        for elem in remaining_nodes:
+            temp_paths = paths[i].copy()
+            temp_paths[len(temp_paths)] = elem
+            temp_arr.append(temp_paths)
+
+        new_paths.extend(recursively_find_pd_paths(G, X, temp_arr, Y))
+
+    return new_paths
+
+
+def possibly_directed_path(G, X: Optional[Set] = None, Y: Optional[Set] = None):
+
+    if isinstance(X, set):
+        x_neighbors = get_X_neighbors(G, X)
+    else:
+        nbr_temp = G.neighbors(X)
+        nbr_possible = check_back_arrow(nbr_temp)
+        x_neighbors = []
+
+        for elem in nbr_possible:
+            temp = dict()
+            temp[0] = X
+            temp[1] = elem
+            x_neighbors.append(temp)
```

Comment:
I'm confused why there's two paths depending on the type of X. Can't we just use `{G.neighbors(x) for x in X}` to get a set of all neighbors of X if X is a set?

Reply:
I did not know that G.neighbors(x) can take a set as an argument as well. Thanks for pointing this out.

Reply:
I tried running G.neighbors on a set of nodes and it errors out:

`TypeError: unhashable type: 'set'`

Is it supposed to throw this error or is it a bug?

Reply:
Ah yeah that is correct. I meant can we just have a one liner that iterates over the nodes in X and adds to a nbr set? The resulting nbr set is just all the nodes that are nbrs of X.

Or is there anything more nuanced happening?

Reply:
I wrote that function to account for the fact that the neighbor of a member of X might be another member of X. Given the following definition of a proper path:
>  A path from X to Y is proper (wrt X) if only its first node is in X

Am I wrong? Or do we need to to do this check?